CommonTools/ParticleFlow/plugins/PFMuonUntagger.cc

```diff
@@ -0,0 +1,121 @@
+    public:
+        PFMuonUntagger(const edm::ParameterSet&);
+        ~PFMuonUntagger() {};
+
+        void produce(edm::Event&, const edm::EventSetup&);
+
+    private:
+        edm::EDGetTokenT<std::vector<reco::Muon> > muons_;
+        std::vector<edm::EDGetTokenT<reco::CandidateView>> badmuons_;
+
+        template<typename H1, typename H2>
+        void writeAssociation(edm::Event &out, const H1 &from, const H2 &to, const std::vector<int> indices, const std::string &name) {
+            typedef edm::Association<std::vector<reco::Muon>> AssoMap;
+            std::unique_ptr<AssoMap> assomap(new AssoMap(to));
+            typename AssoMap::Filler filler(*assomap);
+            filler.insert(from, indices.begin(), indices.end());
+            filler.fill();
+            out.put(std::move(assomap), name);
+        }
+
+        template<typename H1>
+        void writeValueMap(edm::Event &out, const H1 &from, const std::vector<int> values, const std::string &name) {
+            typedef edm::ValueMap<int> IntMap;
+            std::unique_ptr<IntMap> intmap(new IntMap());
+            typename IntMap::Filler filler(*intmap);
+            filler.insert(from, values.begin(), values.end());
+            filler.fill();
+            out.put(std::move(intmap), name);
+        }
+
+};
+
+PFMuonUntagger::PFMuonUntagger(const edm::ParameterSet &iConfig) :
+    muons_(consumes<std::vector<reco::Muon>>(iConfig.getParameter<edm::InputTag>("muons")))
+{
+    for (const auto & src : iConfig.getParameter<std::vector<edm::InputTag>>("badmuons")) {
+        badmuons_.push_back(consumes<reco::CandidateView>(src));
+    }
+
+    produces<std::vector<reco::Muon>>();
+    produces<edm::ValueMap<int>>("oldPF");
+    produces<edm::Association<std::vector<reco::Muon>>>("newToOld");
+    produces<edm::Association<std::vector<reco::Muon>>>("oldToNew");
+    produces<std::vector<reco::Muon>>("bad");
+    produces<edm::Association<std::vector<reco::Muon>>>("badToNew");
+    produces<edm::Association<std::vector<reco::Muon>>>("newToBad");
+}
+
+void PFMuonUntagger::produce(edm::Event &iEvent, const edm::EventSetup&)
+{
+    edm::Handle<std::vector<reco::Muon>> muons;
+    iEvent.getByToken(muons_, muons);
+
+    unsigned int n = muons->size();
+    std::unique_ptr<std::vector<reco::Muon>> copy(new std::vector<reco::Muon>(*muons));
+    std::vector<int> oldPF(n); 
+    std::vector<int> dummyIndices(n); 
+    for (unsigned int i = 0; i < n; ++i) {
+        oldPF[i] = (*copy)[i].isPFMuon();
+        dummyIndices[i] = i;
+    }
+
+    edm::Handle<reco::CandidateView> badmuons;
+    for (const auto & tag : badmuons_) {
+        iEvent.getByToken(tag, badmuons);
+        for (unsigned int j = 0, nj = badmuons->size(); j < nj; ++nj) {
+            reco::CandidatePtr p = badmuons->ptrAt(j);
+            if (p.isNonnull() && p.id() == muons.id()) {
+                reco::Muon &mu = (*copy)[p.key()];
+                mu.setType(mu.type() & ~reco::Muon::PFMuon);
```

Comment:
I suppose, we do not care about pfP4 and other PF-related values in this muon.
Or is it preserved to know what it was originally?